File: bench/promptgreen/tasks_pending/t23/hidden_tests.mjs
// Hidden test for t23 (prime count sieve). Never shown to an author.
const EXPECTED = '168\n';

export async function run(compileFn, source) {
  const r = compileFn(source);
  const checks = [];

  const basic = r.ok === true && r.stdout === EXPECTED;
  checks.push(basic);

  // Edge: output must be exactly one line, no stray prints.
  const singleLine = r.ok === true && r.stdout.split('\n').filter(x => x.length > 0).length === 1;
  checks.push(singleLine);

  // Edge: 997 is the largest prime below 1000; a sieve that mis-handles the
  // upper boundary (e.g. off-by-one on N or on the inner marking loop) will
  // not land on exactly 168.
  const exactCount = r.ok === true && r.stdout.trim() === '168';
  checks.push(exactCount);

  const green = checks.every(Boolean);
  return { green, detail: { expected: EXPECTED, got: r.stdout, ok: r.ok, checks } };
}

File: bench/promptgreen/tasks_pending/t23/py/hidden_tests.py
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REF = os.path.join(HERE, "reference.py")


def run_ref():
    result = subprocess.run([sys.executable, REF], capture_output=True, text=True)
    assert result.returncode == 0, f"reference.py exited nonzero: {result.stderr}"
    return result.stdout


def main():
    out = run_ref()
    lines = [l for l in out.split("\n") if l.strip() != ""]

    assert len(lines) == 1, f"expected exactly one output line, got {lines}"
    assert lines[0].strip() == "168", f"expected 168 primes below 1000, got {lines[0]}"

    # Edge: 997 (largest prime below 1000) must independently verify as prime
    # via trial division, confirming the sieve's boundary handling.
    def is_prime(x):
        if x < 2:
            return False
        d = 2
        while d * d <= x:
            if x % d == 0:
                return False
            d += 1
        return True

    assert is_prime(997), "sanity check: 997 must be prime"

    print("OK")


if __name__ == "__main__":
    main()
